src/v/cloud_storage/remote_partition.cc

```diff
@@ -190,16 +191,19 @@ class partition_record_batch_reader_impl final
               "gate_closed_exception while reading from remote_partition");
             _it = _end;
             _reader = {};
-        } catch (const std::exception& e) {
+        } catch (const std::bad_alloc& e) {
             vlog(
               _ctxlog.warn,
-              "exception thrown while reading from remote_partition: {}",
+              "bad_alloc thrown while reading from remote_partition: {}",
               e.what());
-            _it = _end;
-            _reader = {};
```

Comment:
I wonder if calling `remote_partition::evict_reader` instead of simply resetting it would solve the problem

Reply:
Although synchronously closing it with `partition_record_batch_reader_impl::set_end_of_stream` looks okay too and more straightforward.

Reply:
I'm not sure it would. If the reader is returned and handed out before the eviction takes place, the original issue can still happen. Pretty unlikely though.